Encourage not using the static ResourcesPlugin.getWorkspace()

laeubi opened the pull request:
https://github.com/eclipse-platform/eclipse.platform.resources/issues/52

ruspl-afed commented on bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java:
```diff
@@ -416,12 +417,16 @@ public static ResourcesPlugin getPlugin() {
+	 * @return the workspace that was created by the single instance of this plug-in
+	 *         class.
 	 */
+	@Deprecated(forRemoval = true)
```
In case of removal it should be a major version change, perhaps for the whole Eclipse SDK, since the most  part of Eclipse plug-ins that use IResource will stop work.

laeubi replied:
Yep, that's why I just marked it for removal so if we ever do a major upgrade...

iloveeclipse replied:
Please change to "forRemoval = false" and that could be discussed (after fixing all the new warnings in resources code and of others agree).

"forRemoval = true" is not an option, sorry, see comments from Ed and others.

laeubi replied:
I don't get why this makes any difference here but anyways I'll remove the "offending" flag then, this literally makes no sense to me as a major version change would require all plugins to become invalid anyways...

merks commented:
I know you won't like it, but of course everyone is using this and have been for 20 years.   I use this and I have stuff that runs in Helios.  Everyone using it will not want to change it to something else because they don't have the problem you describe.  The deprecation doesn't even show a concrete replacement but rather alludes to some vague ideas.  So as a consumer of this API my reaction will be, oh god, not more disruptive changes!  What should I do?  Why is this important for me to change?  I don't have this problem. Why don't these developers focus on the 140 bugs that are open instead of creating work for me with disruptive changes?  Maybe I should find a new technology base because I can no longer rely on being able to move to new releases of this technology without a major rework 4 times a years...

I have this in the EMF API:
```
  public static IWorkspaceRoot getWorkspaceRoot()
  {
    if (workspaceRoot == null && IS_RESOURCES_BUNDLE_AVAILABLE && System.getProperty("org.eclipse.emf.ecore.plugin.EcorePlugin.doNotLoadResourcesPlugin") == null)
    {
      workspaceRoot = ResourcesPlugin.getWorkspace().getRoot();
    }
    return workspaceRoot;
  }
```
I wouldn't know what to do with that.  I don't want to break my API.  Please don't make me do that.

If you feel compelled to make busy work for me and for the entire Eclipse ecosystem, I suggest you start bottom up and look at the call hierarchy of that method in the SDK workspace and start removing the ~1,700 uses of this method.   

![image](https://user-images.githubusercontent.com/208716/163528957-2d2c13f4-e9c4-4a00-8a73-2815a2f7482e.png)


Then you can suggest concrete replacements and you can socialize this idea with the community that is impacted by it.  Consider too how many uses exist in the universe and how much work this universe will expend on this idea and to what gain; they'll be busy with this, and not development new features or fixing bugs...

laeubi commented:
> The deprecation doesn't even show a concrete replacement but rather alludes to some vague ideas

Sorry to say but Eclipse / Platform is heavily based on OSGi and plugins, so anyone that finds "Service" vague should read some basic tutorials about OSGi and start using the technique he likes to contribute to.

> Why is this important for me to change?

I have described some of the most common things about static singleton you can read more here: https://stackoverflow.com/questions/137975/what-are-drawbacks-or-disadvantages-of-singleton-pattern

> Then you can suggest concrete replacements and you can socialize this idea with the community that is impacted by it.

The service for this was introduced in 2009 and still people are rely on it because it is

1. recommended in the docs
2. they probably don't care

so it is time to change here and anyone can support stuff dating back 13 years... If you like static access you can simply track the service in your bundle-activator:
```
public class MyActivator implements BundleActivator {
  private static MyActivator theUsualStaticStuff;

  private ServiceTracker<IWorkspace, IWorkspace> workspaceTracker;

  @Override
  public void start(BundleContext context) throws Exception {
    this.workspaceTracker = new ServiceTracker<>(context, IWorkspace.class, null);
    this.workspaceTracker.open();
    theUsualStaticStuff = this;

  }

  @Override
  public void stop(BundleContext context) throws Exception {
    theUsualStaticStuff = null;
  }

  public static IWorkspace getWorkspaceTheStaticWay() {
    return Objects.requireNonNull(theUsualStaticStuff, "seems this plugin is not started").workspaceTracker
        .getService();
  }

}
```

laeubi commented:
> I suggest you start bottom up and look at the call hierarchy of that method in the SDK workspace and start removing the ~1,700 uses of this method.

I'm already removing calls to this method in the resources plugin, and in almost all cases it is just used even if the `IWorkspace` is already available without any special effort. And as long as we do not strongly discourage the access this will never change as it hasn't changed in 13 year.

ruspl-afed commented on bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/CheckMissingNaturesListener.java:
```diff
@@ -32,6 +33,11 @@ public class CheckMissingNaturesListener implements IResourceChangeListener, IPr
 
 	public static final String MARKER_TYPE = ResourcesPlugin.getPlugin().getBundle().getSymbolicName() + ".unknownNature"; //$NON-NLS-1$
 	public static final String NATURE_ID_ATTRIBUTE = "natureId"; //$NON-NLS-1$
+	private Workspace workspace;
```
final?

laeubi commented:
I now replaced the deprecation with a kind reminder in the java-doc, I hope this would not be considered to disruptive.

ruspl-afed commented on bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/utils/Messages.java:
```diff
@@ -295,6 +295,7 @@ public class Messages extends NLS {
 	public static String resources_updating;
 	public static String resources_updatingEncoding;
 	public static String resources_workspaceClosed;
```
Do we still use this message in other places?

laeubi replied:
Yes it is used in org.eclipse.core.internal.resources.Workspace.prepareOperation(ISchedulingRule, IProgressMonitor)

laeubi commented:
> I think the documentation should give more specific concrete code examples of how to get the workspace

The problem is: Ther is not one way and I think we can't explain general coding guidelines in the javadoc.

Just some examples:

- An activator might simply use a `ServiceTracker`
- A DS component could simply use `@Refeference`
- If you have a `IResource` at hand you can simply call `IResource.getWorkspace()`
- One might even use `ServiceCaller` or other helpers
- One even might use Spring-DM or Blueprint
- ...

beside that there are million of examples/ tutorial/... on how to fetch services with just simply searching for "fetch osgi service"

mickaelistria commented:
Documentatin is the key and documentation follows UX laws: people want to go fast and are lazy and won't disturb there current work too much because they have their own goals that are more important to them than what's best for the framework; so really if you want to influence people to not use this API, we need to be more helpful and give maybe the 1 or 2 main examples. Javadoc may not be the best place for it in theory, but in practice it's the only thing users might look at.
Good old singleton works very well; so we need something really easy and accessible if we want people to move away from it; otherwise they'll just not adopt newer ways.

Also please add IWorkspace to https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fservices , and maybe enrich this page with some examples of how to get those services.

ruspl-afed reviewed (approved):
Now it looks good to me.
Thank you for your effort @laeubi and for your bold step forward to start this painful but necessary discussion.

laeubi commented:
> people want to go fast and are lazy and won't disturb there current work too much because they have their own goals

Yeah but I'm also lazy and don't feel to have the time supporting others laziness ;-)

Anyways I updated `IResourceChangeListener` to make people more aware of its usage without even the need to fetch the `IWorkspace` at all and moved the example from the (internal) implementing class to the interface description.

@iloveeclipse please consider approving this request if you think it is fine now, merging is currently blocked by your review.

mickaelistria commented:
> Yeah but I'm also lazy and don't feel to have the time supporting others laziness ;-)

Welcome to world of developing a platform or a framework.

> Anyways I updated IResourceChangeListener to make people more aware of its usage without even the need to fetch the IWorkspace at all and moved the example from the (internal) implementing class to the interface description.

That's good. Could be a separate commit if you prefer.

I'll also review with a -1 as I really believe that the change in `ResourcesPlugin.getWorkspace()` is pointless without more concrete example.

mickaelistria commented on bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java:
```diff
@@ -416,12 +417,16 @@ public static ResourcesPlugin getPlugin() {
 	 *
-	 * @return the workspace that was created by the single instance of this
-	 *   plug-in class.
+	 * <b>Hint:</b> Accessing the Workspace in a static way is prone to start-up
+	 * order problem, please consider using any of your favorite techniques
+	 * (Declarative Services, ServiceTracker, Blueprint, ...) instead.
```
This needs some examples as code snippet.

laeubi replied:
I have added some code sample to the top of the javadoc now.

mickaelistria commented on bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java:
```diff
@@ -32,12 +33,43 @@
+ * cause this plug-in to be activated and the workspace (if any) to be loaded
+ * from disk and initialized.
+ *
+ * <p>
+ * Fetching the service could be performed in several ways:
+ * <ul>
+ * <li>For example in a Bundle/Plugin Activator:
+ *
+ * <pre>
+ * ServiceTracker<IWorkspace, IWorkspace> workspaceTracker = new ServiceTracker<>(bundleContext, IWorkspace.class,
+ * 		null);
+ * workspaceTracker.open();
+ * IWorkspace workspace = workspaceTracker.getService();
+ * </pre>
+ *
+ * </li>
+ * <li>Using declarative service with annotations:
+ *
+ * <pre>
+ * &#64;Reference
+ * private IWorkspace workspace;
```
Why not on `getWorksapce` method documentation, since it's the method we aim at replacing? When this gets deprecated, it's the `getWorkpsace()` javadoc consumer will read, not `ResourcesPlugin`'s one.

laeubi replied:
Because the ResourcesPlugin claims it is the way to acquire a workspace, actually I would aim to replace *all* calls to `ResourcesPlugin` and don't want to document each and every method with an example. For that reason as mentioned before I do not think adding "how to use osgi" examples at all as everyone will consider the documentation at different levels (the more interesting question would be why not place this at the `IWorkspace`....)

mickaelistria replied:
We can slightly change the doc of `ResourcesPlugin`, but I believe doc on alternative to `getWorkspace` would still fit much better on `getWorkspace`

laeubi replied:
I moved the example to `IWorkspace` and added a hint to `getWorkspace()`

mickaelistria reviewed (approved):
I'm fine with the documentation in its current form. thanks.

laeubi commented:
I split the part with the IResourceChangeListener now into a separate PR: https://github.com/eclipse-platform/eclipse.platform.resources/pull/54

laeubi commented:
Builds are clean and we have two approvals so its up to you @iloveeclipse for a final review.

iloveeclipse commented:
Please fix new javadocs added, we have releng errors reported :

https://download.eclipse.org/eclipse/downloads/drops4/I20220416-1800/compilelogs/platform.doc.isv.javadoc.txt
